[llvm][TableGen] Add a README to the main TableGen folder

DavidSpickett opened the pull request:
Though I doubt that many people will land here directly, I thought it odd that we didn't have one.

The intro I've copied from the programmer's reference and added a simple example. Then some links to resources and tools, which is the main reason to have this page.

Commit:
[llvm][TableGen] Add a README to the main TableGen folder

Though I doubt that many people will land here directly, I
thought it odd that we didn't have one.

The intro I've copied from the programmer's reference and
added a simple example. Then some links to resources and tools,
which is the main reason to have this page.

DavidSpickett commented:
@joker-eph do you know if there's any other MLIR tablegen things worth mentioning? I've just linked to the LSP server for now. Maybe MLIR has a better page than that I should link instead.

joker-eph commented:
We have our ODS definition docs: https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/  and https://mlir.llvm.org/docs/DefiningDialects/Operations/

This isn't generic "TableGen" but your "Resources for learning the language" aren't all generic either I believe (by the way should this be curated and split into "TableGen" vs "some TableGen backends"?)

Commit:
* Split language and backend sections
* Add MLIR links

DavidSpickett commented:
Added the MLIR links and split language and backend resources.

TableGen is live in Compiler Explorer now, but I haven't added a "trunk" variant yet. So I'll add links for that in a follow up when that's there.

mshockwave commented on llvm/utils/TableGen/README.md:
````diff
@@ -0,0 +1,54 @@
+
+```
+class Hello <string _msg> {
+	string msg = !strconcat("Hello ", _msg);
````
nit: why is the indentation here different from that of line 19 and 23?

DavidSpickett replied:
I wrote it in Compiler Explorer where <tab> is an actual tab, llvm-tblgen outputs 2 space tabs I presume. I've fixed it.

mshockwave reviewed (approved):
This is pretty neat! LGTM w/ a minor comment.

> TableGen is live in Compiler Explorer now, but I haven't added a "trunk" variant yet. So I'll add links for that in a follow up when that's there.

Please do! CE is easily approachable for most people.

Commit:
Fix indentation in example
